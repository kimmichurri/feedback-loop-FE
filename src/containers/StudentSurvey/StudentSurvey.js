import React, { Component } from 'react'
import Response from '../Response/Response'
import { connect } from 'react-redux'
import { handlePost } from '../../thunks/handlePost'
import { handleGet } from '../../thunks/handleGet'
import { setStudentSurveys } from '../../actions'
import cogoToast from 'cogo-toast'

export class StudentSurvey extends Component {
  constructor(props) {
    super(props);
    this.state = {
      allResponses: [],
      membersReviewed: 0,
    }
  }

  componentDidMount() {
    const path = this.props.location.pathname
    const splitPath = path.split('/')
    const id = splitPath[splitPath.length -1]
    const storedSurveys = this.props.studentSurveys
    const foundSurvey = storedSurveys.find(survey => {
      return survey.id == id
    })
    this.setState({
      surveyName: foundSurvey.surveyName,
      id: foundSurvey.id,
      questions: foundSurvey.questions,
      members: foundSurvey.groups[0].members,
    })
  }

  renderResponse = () => {
    const { members, questions } = this.state
    return members.map(member => {
      return <Response key={member.id} member={member} questions={questions} collectResponses={this.collectResponses}/>
    })
  }

  collectResponses = (individualResponse) => {
    if(this.state.allResponses.length >= 1) {
      this.setState({
        allResponses: [...this.state.allResponses, ...individualResponse],
        membersReviewed: this.state.membersReviewed + 1
      })
    } else {
      this.setState({
        allResponses: [...individualResponse],
        membersReviewed: this.state.membersReviewed + 1
      })
    }
  }

  postResponse = async () => {
    if(this.state.membersReviewed < this.state.members.length) {
      this.warnToast('Please complete surveys before submitting your response')
    } else {
      const url = "https://turing-feedback-api.herokuapp.com/api/v1/responses"
      const options = {
        method: 'POST',
        body: JSON.stringify({
          api_key: this.props.user,
          responses: this.state.allResponses
        }),
        headers: {
          'Content-Type': 'application/json'
        }
      }
      this.props.handlePost(url, options)
      this.handleSuccess('Thank you for completing this survey')
    }
  }

  warnToast = (message) => {
    cogoToast.warn(message, {position: 'bottom-left'})
  }

  handleSuccess = async (message) => {
    cogoToast.success(message, {position: 'bottom-left'})
    const url = `https://turing-feedback-api.herokuapp.com/api/v1/surveys/pending?api_key=${this.props.user}`
    const surveys = await this.props.handleGet(url)
    this.props.setStudentSurveys(surveys)
    this.props.history.push('/student-dashboard')
  }

  render() {
    return(
      <div className='student-survey'>
        <p className='response-survey-name'>{this.state.surveyName}</p>
        {this.state.members && this.renderResponse()}
        <button className='response-button' onClick={this.postResponse}>Submit Response</button>
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  studentSurveys: state.studentSurveys,
  user: state.user
})

export const mapDispatchToProps = (dispatch) => ({
  handlePost: (url, options) => dispatch(handlePost(url, options)),
  handleGet: (url) => dispatch(handleGet(url)),
  setStudentSurveys: (surveys) => dispatch(setStudentSurveys(surveys))
})

export default connect(mapStateToProps, mapDispatchToProps)(StudentSurvey)