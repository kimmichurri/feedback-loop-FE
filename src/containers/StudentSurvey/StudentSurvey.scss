.student-survey {
  background: #fff;
  width: 80%;
  margin: auto;
  padding: 25px;
  height: 500px;
  display: grid;
  grid-template-rows: 100px 1fr 50px;
  overflow: auto;
}

.member-survey {
  border: solid #f4f4f4 .5px;
  box-shadow: 0 3px 5px rgba(0,0,0,0.20);
  padding: 10px;
  margin-bottom: 10px;
}

.all-responses {
  height: 40vh;
  overflow: scroll;
}

.response-card {
  display: flex;
  flex-direction: column;
  text-align: left;
  padding: 30px 10px;
  background: $white;
  border: solid #f4f4f4 .5px;
  box-shadow: 0 3px 5px rgba(0,0,0,0.20);
  margin-bottom: 5px;
}

.question-title {
  width: 100%;
  border: none;
  border-bottom: solid $main-grey 1px;
  font-size: 20px;
  color: $accent-charcoal;
}

.response-container {
  display: flex;
  flex-direction: column;
}

.response-option {
  display: flex;
  align-items: center;
}

.response-description {
  padding-left: 10px;
  font-size: 15px;
}

.response-button {
  background-color: $main-blue;
  border-radius: 10px;
  border: none;
  color: $accent-teal;
  font-family: $main-font;
  font-weight: bold;
  font-size: 16px;
  height: 50px;
  width: 100%;
  margin: 10px 0px;
  &:hover {
    cursor: pointer;
    background-color: $main-blue-hover;
  }
  &:focus {
    outline: none;
  }
  &:disabled {
    background-color: $light-grey;
    cursor: not-allowed;
  }
}

.member-button {
  border: none;
  font-weight: normal;
  font-size: 16px;
  height: 50px;
  width: 100%;
  margin: 10px 0px;
  &:focus {
    outline: none;
  }
}

.close-button-container {
  display: flex;
  justify-content: flex-end;
}

.close-button {
  background-color: $main-blue;
  border-radius: 10px;
  border: none;
  color: $accent-teal;
  font-family: $main-font;
  font-weight: bold;
  font-size: 16px;
  height: 50px;
  width: 35%;
  margin: 10px 0px;
  &:hover {
    cursor: pointer;
    background-color: $main-blue-hover;
  }
  &:focus {
    outline: none;
  }
}

.response-survey-name {
  font-size: 34px;
  height: 51px;
  width: 100%;
  border: none;
  margin: 0px;
  color: $main-grey;
  line-height: 135%;
  font-weight: normal;
  padding: 25px 0px;
  text-align: left;
}
