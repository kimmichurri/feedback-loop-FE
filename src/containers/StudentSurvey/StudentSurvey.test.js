import React from 'react'
import { shallow } from 'enzyme'
import { StudentSurvey, mapStateToProps, mapDispatchToProps } from './StudentSurvey'
jest.mock('../../thunks/handlePost')

describe('StudentSurvey', () => {
  let wrapper
  let mockUser
  let mockHandlePost
  let mockStudentSurveys
  let mocklocation

  beforeEach(() => {
    mockUser = '12345'
    mockStudentSurveys = [
      { 
        id: 1, 
        surveyName: "Survey 1", 
        groups: [{members: [{id: 1, name: "name1"}, {id: 2, name: "name2"}]}]},
    ]
    mockHandlePost = jest.fn()
    mocklocation = { pathname : 'mockUrl/1' }
  
    wrapper = shallow(
      <StudentSurvey
        studentSurveys={mockStudentSurveys}
        user={mockUser}
        handlePost={mockHandlePost}
        location={mocklocation}
      />
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('mapStateToProps', () => {
    it('should return the expected state as props', () => {
      const mockState = {
        user: '12345',
        studentSurveys: [
          { id: 1, name: "Survey 1"},
          { id: 2, name: "Survey 2"}
        ],
        fakeState: 'fakeState'
      }
      const expected = {
        user: '12345',
        studentSurveys: [
          { id: 1, name: "Survey 1"},
          { id: 2, name: "Survey 2"}
        ]
      }
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should return handlePost to dispatch', () => {
      const mockUrl = 'www.postresponses.com'
      const mockOptions = {
        method: 'POST',
        body: { api_key: "12345",
                responses: [
                  {
                    "question": 7,
                    "answer": 11,
                    "member": 2
                  },
                  {
                    "question": 7,
                    "answer": 10,
                    "member": 6
                  },
                ]
              },
        headers: {
          'Content-Type': 'application/json'
        }
      }
      const mockDispatch = jest.fn()
      const handlePost = jest.fn()
      const actionToDispatch = handlePost(mockUrl, mockOptions)
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.handlePost(mockUrl, mockOptions)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })

})