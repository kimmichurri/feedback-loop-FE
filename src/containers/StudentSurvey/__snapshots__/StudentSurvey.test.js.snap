// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentSurvey should match the snapshot 1`] = `
<div
  className="student-survey"
>
  <p
    className="response-survey-name"
  >
    Survey 1
  </p>
  <Response
    collectResponses={[Function]}
    key="1"
    member={
      Object {
        "id": 1,
        "name": "name1",
      }
    }
  />
  <Response
    collectResponses={[Function]}
    key="2"
    member={
      Object {
        "id": 2,
        "name": "name2",
      }
    }
  />
  <button
    className="response-button"
    onClick={[Function]}
  >
    Submit Response
  </button>
</div>
`;
